ocaml-lsp-server/test/e2e-new/type_enclosing.ml

```diff
@@ -0,0 +1,241 @@
+end
+
+let f () = (M.A : M.t)
+
+let _ = M.A
+
+module N = struct
+  type t = A of int
+  let x = 3
+end
+
+let _ = Some (N.A 3)
+
+let _ = N.x
+|}
+
+let%expect_test "type enclosing constructors_and_path - 1" =
+  let source = cons_ml in
+  let request client =
+    let open Fiber.O in
+    let position = Position.create ~line:3 ~character:13 in
+    let index = 0 in
+    let+ response = call_type_enclosing client position index in
+    print_type_enclosing response
+  in
+  Helpers.test source request;
+  [%expect
+    {|
+    {
+      "index": 0,
+      "enclosings": [
+        {
+          "end": { "character": 14, "line": 3 },
+          "start": { "character": 13, "line": 3 }
+        }
+      ],
+      "type": "t"
+    } |}]
+
+let%expect_test "type enclosing constructors_and_path - 2" =
+  let source = cons_ml in
+  let request client =
+    let open Fiber.O in
+    let position = Position.create ~line:7 ~character:4 in
+    let index = 0 in
+    let+ response = call_type_enclosing client position index in
+    print_type_enclosing response
+  in
+  Helpers.test source request;
+  [%expect
+    {|
+    {
+      "index": 0,
+      "enclosings": [
+        {
+          "end": { "character": 5, "line": 7 },
+          "start": { "character": 4, "line": 7 }
+        },
+        {
+          "end": { "character": 11, "line": 7 },
+          "start": { "character": 2, "line": 6 }
+        },
+        {
+          "end": { "character": 11, "line": 7 },
+          "start": { "character": 6, "line": 5 }
+        }
+      ],
+      "type": "t"
+    } |}]
+
+let%expect_test "type enclosing constructors_and_path - 3" =
+  let source = cons_ml in
+  let request client =
+    let open Fiber.O in
+    let position = Position.create ~line:16 ~character:8 in
+    let index = 0 in
+    let+ response = call_type_enclosing client position index in
+    print_type_enclosing response
+  in
+  Helpers.test source request;
+  [%expect
+    {|
+    {
+      "index": 0,
+      "enclosings": [
+        {
+          "end": { "character": 11, "line": 16 },
+          "start": { "character": 8, "line": 16 }
+        }
+      ],
+      "type": "sig type t = A type u = A | B end"
+    } |}]
+
+let%expect_test "type enclosing constructors_and_path with reconstruction - 4" =
+  let source = cons_ml in
+  let request client =
+    let open Fiber.O in
+    let position = Position.create ~line:23 ~character:14 in
+    let index = 0 in
+    let+ response = call_type_enclosing client position index in
+    print_type_enclosing response
+  in
+  Helpers.test source request;
+  [%expect
+    {|
+    {
+      "index": 0,
+      "enclosings": [
+        {
+          "end": { "character": 20, "line": 23 },
+          "start": { "character": 13, "line": 23 }
+        },
+        {
+          "end": { "character": 20, "line": 23 },
+          "start": { "character": 8, "line": 23 }
+        }
+      ],
+      "type": "sig type t = A of int val x : int end"
+    } |}]
```

Comment:
We should also put a focus on the `range_end` cases that rely on tricky position comparisons and index shifting :-)